salt/pillar/consul_pillar.py

```diff
@@ -0,0 +1,204 @@
+appears after the shared root:
+
+.. code-block:: yaml
+
+    ext_pillar:
+      - consul: my_consul_config root=/salt-shared
+      - consul: my_other_consul_config root=/salt-private/%(minion_id)s
+
+'''
+from __future__ import absolute_import
+
+# Import python libs
+import logging
+
+import re
+
+from salt.exceptions import CommandExecutionError
+
+# Import third party libs
+try:
+    import consul
+    HAS_CONSUL = True
+except ImportError:
+    HAS_CONSUL = False
+
+__virtualname__ = 'consul'
+
+# Set up logging
+log = logging.getLogger(__name__)
+
+
+def __virtual__():
+    '''
+    Only return if python-consul is installed
+    '''
+    return __virtualname__ if HAS_CONSUL else False
+
+
+def ext_pillar(minion_id,
+               pillar,  # pylint: disable=W0613
+               conf):
+    '''
+    Check consul for all data
+    '''
+    comps = conf.split()
+
+    profile = None
+    if comps[0]:
+        profile = comps[0]
+    client = get_conn(__opts__, profile)
+
+    path = ''
+    if len(comps) > 1 and comps[1].startswith('root='):
+        path = comps[1].replace('root=', '')
+
+    # put the minion's ID in the path if necessary
+    path %= {
+        'minion_id': minion_id
+    }
+
+    try:
+        pillar = fetch_tree(client, path)
+    except KeyError:
+        log.error('No such key in consul profile {0}: {1}'
+                  .format(profile, path))
+        pillar = {}
+
+    return pillar
+
+
+def consul_fetch(client, path):
+    '''
+    Query consul for all keys/values within base path
+    '''
+    return client.kv.get(path, recurse=True)
+
+
+def fetch_tree(client, path):
+    '''
+    Grab data from consul, trim base path and remove any keys which
+    are folders. Take the remaining data and send it to be formatted
+    in such a way as to be used as pillar data.
+    '''
+    index, items = consul_fetch(client, path)
+    ret = {}
+    has_children = re.compile(r'/$')
+
+    log.debug('Fetched items: %r', format(items))
+
+    for item in reversed(items):
+        key = re.sub(r'^' + path + '/?', '', item['Key'])
+        if key != "":
+            log.debug('key/path - {0}: {1}'.format(path, key))
+            log.debug('has_children? %r', format(has_children.search(key)))
+        if has_children.search(key) is None:
+            ret = pillar_format(ret, key.split('/'), item['Value'])
+            log.debug('Fetching subkeys for key: %r', format(item))
+
+    return ret
+
+
+def pillar_format(ret, keys, value):
+    '''
+    Perform data formatting to be used as pillar data and
+    merge it with the current pillar data
+    '''
+    if value is None:
+        return ret
+    array_data = value.split('\n')
+    pillar_value = array_data[0] if len(array_data) == 1 else array_data
+    keyvalue = keys.pop()
+    pil = {keyvalue: pillar_value}
+    keys.reverse()
+    for k in keys:
+        pil = {k: pil}
+
+    return dict_merge(ret, pil)
+
+
+def dict_merge(d1, d2):
```

Comment:
can you make use of https://github.com/saltstack/salt/blob/develop/salt/utils/dictupdate.py?